Make sure fixedin field is lower than 256 chars

mattiaverga opened the pull request:
Make sure fixedin field is lower than 256 chars (fixes #1430 ).

Instead of adding the new version to the fixedin list and then check the length of the resultant string, check if the new version can be added to the string without going over the 255 chars limit.

mattiaverga committed:
Make sure fixedin field is lower than 256 chars.
fixes #1430

Check it's lower than 255 not considering space

Signed-off-by: Mattia Verga <mattia.verga@tiscali.it>

bowlofeggs commented:
jenkies test

bowlofeggs commented on bodhi/server/bugs.py:
```diff
@@ -201,15 +201,17 @@ def close(self, bug_id, versions, comment):
+                # Add our build if its not already there
+                # but only if resultant string length is lower than 256 chars
+                if (version not in fixedin) and (len(fixedinString)+len(version) < 255):
+                    args['fixedin'] = fixedinString + " " + version
```
This will prepend a space if ```fixedinString``` is the empty string, which will happen on the first item that adds a fixedin to a bug. I suggest something like this instead:

```
args['fixedin'] = ' '.join([fixedin_str, version])
```

bowlofeggs commented on bodhi/server/bugs.py:
```diff
@@ -201,15 +201,17 @@ def close(self, bug_id, versions, comment):
                 # - the full NVR as it appears in koji
                 # - space-separated if there's more than one.
-                args['fixedin'] = " ".join(fixedin)
+                fixedinString = " ".join(fixedin)
```
For [PEP-8](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names), this should be named ```fixedin_str``` (or ```fixedin_string``` if you prefer the full name).

bowlofeggs commented on bodhi/server/bugs.py:
```diff
@@ -201,15 +201,17 @@ def close(self, bug_id, versions, comment):
+                
+                # Add our build if its not already there
+                # but only if resultant string length is lower than 256 chars
+                if (version not in fixedin) and (len(fixedinString)+len(version) < 255):
```
Also for [PEP-8](https://www.python.org/dev/peps/pep-0008/#other-recommendations), there should be spaces surrounding the ```+``` operator.

bowlofeggs replied:
```flake8``` mentioned this too in the CI test run:

```
rawhide	./bodhi/server/bugs.py:213:68: E226 missing whitespace around arithmetic operator
rawhide	                if (version not in fixedin) and (len(fixedinString)+len(version) < 255):
rawhide	     
```

bowlofeggs commented on bodhi/server/bugs.py:
```diff
@@ -201,15 +201,17 @@ def close(self, bug_id, versions, comment):
+                fixedinString = " ".join(fixedin)
+                
+                # Add our build if its not already there
+                # but only if resultant string length is lower than 256 chars
```
It might be good to put a link to the bug report in this comment block so future readers know where the 256 limit comes from.

bowlofeggs reviewed (changes requested):
Hello, and thanks for the patch! I've left a few notes of changes we'll want to make. In addition to those changes:

* We will want to add tests for this change to ```bodhi/tests/server/test_bugs.py```. I suggest a test that makes sure we don't exceed the 255 limit, and also a test to ensure we don't prepend an empty space on the first fixedin version.
* Please squash your two commits into one commit.
* Please move the ```fixes #1430``` onto its own line in your commit message. This way GitHub will automatically close the bug when this gets merged, which is handy ☺
* Please sign the [DCO](https://developercertificate.org/) in your commit message. The way to do that is described in Bodhi's [contribution guidelines](https://bodhi.fedoraproject.org/docs/developer/index.html#contribution-guidelines).

bowlofeggs commented on bodhi/server/bugs.py:
```diff
@@ -201,15 +201,17 @@ def close(self, bug_id, versions, comment):
                 # - space-separated if there's more than one.
-                args['fixedin'] = " ".join(fixedin)
+                fixedinString = " ".join(fixedin)
+                
```
```flake8``` fails on this line because it contains whitespace:

```
rawhide	./bodhi/server/bugs.py:210:1: W293 blank line contains whitespace
rawhide	                
rawhide	^
```

If you are using the Vagrant development environment, you can reproduce this test by running ```btest```. You can also run ```flake8``` directly if you like.

mattiaverga committed:
Avoid starting whitespace + PEP-8 fixes

Signed-off-by: Mattia Verga <mattia.verga@tiscali.it>

Fix whitespaces on empty line

Signed-off-by: Mattia Verga <mattia.verga@tiscali.it>

mattiaverga commented:
I'm having some problems creating the tests: I have to figure out how to access the "fixedin" field of the bug.
I tried to test with 
`self.assertEqual(bz._bz.getbug.return_value.fixedin, 'bodhi-3.1.0-1.fc27')`
but it seems that it's not so easy...

mattiaverga commented:
While adding new tests I found that the close() method was only tested for failing, not for success.
So, I wrote a simple test case starting from the existent one, tested against develop and I found that is failing:
```
    @mock.patch('bodhi.server.bugs.log.exception')
    def test_close_successful(self, exception):
        """Test the close() method with a success case."""
        bz = bugs.Bugzilla()
        bz._bz = mock.MagicMock()
        bz._bz.getbug.return_value.component = 'bodhi'

        bz.close(12345, {'bodhi': ['bodhi-3.1.0-1.fc27']}, 'Fixed. Closing bug.')

        bz._bz.getbug.assert_called_once_with(12345)
        bz._bz.getbug.return_value.close.assert_called_once_with('ERRATA',
                                                                comment='Fixed. Closing bug.',
                                                                fixedin='bodhi-3.1.0-1.fc27')
        self.assertEqual(exception.call_count, 0)
```
It fails with:
```
  File "bodhi/server/bugs.py", line 215, in close
    args['fixedin'] = " ".join(fixedin)
TypeError: sequence item 0: expected string, list found
```
In fact, the problem is in the call `bz.close(12345, {'bodhi': ['bodhi-3.1.0-1.fc27']}, 'Fixed. Closing bug.')` that should be `bz.close(12345, {'bodhi': 'bodhi-3.1.0-1.fc27'}, 'Fixed. Closing bug.')`.

So, my question is: is the call in the existent test wrong? Or the versions dict element is supposed to be a list?

mattiaverga commented:
I'm starting to getting lost here...
Looking at models.py, the `versions` object passed to the close() function is:
`versions = {'component' : ['version', ]}`
so in bodhi/server/bugs.py at line 199 the code
`version = versions[bug.component]`
means we have `version = ['version', ]`

After that we do:
```
if version not in fixedin:
    fixedin.append(version)
...
args['fixedin'] = " ".join(fixedin)
```
I can't see how this could work, since we are joining strings with a list of strings. In fact running this code in a console fails just like the testcase I wrote.
BUT it seems that on the real istance of Bodhi this is working without errors... looking at recently closed bugs I can see the fixedin field is populated.

mattiaverga committed:
Add tests for succesfully closing bugs.
Make sure fixed_in field is less than 255 chars and that it doesn't start with whitespace.

Signed-off-by: Mattia Verga <mattia.verga@tiscali.it>

mattiaverga commented:
My fault, the `versions` object is:
versions = {'component' : 'version', }
so it was the call in `test_close_fault` that was wrong, but obviously it was passing because it was triggering an Exception...

Now all the tests pass.

bowlofeggs commented:
jenkies test

bowlofeggs commented:
Hi @mattiaverga! Sorry I've not been very responsive. The past couple of weeks have been really crazy due to the F28 Bodhi activation and I've been working long hours on problems related to that. It sounds like you were able to get some of your questions answered though!

I think CI might be broken right now (also due to some F28 stuff) so Jenkins might say this fails, but I'm running it to see if it's fixed today.

bowlofeggs commented:
I can confirm that the tests pass locally on non-rawhide containers.

bowlofeggs reviewed (approved):
Looks good, thanks for the patch!

I think rawhide should be fixed tomorrow, so I will wait to merge until CI can formally pass, but I think it'll be fine.

bowlofeggs commented:
I opened a PR to create one more test related to this patch: https://github.com/fedora-infra/bodhi/pull/2205

I'll rebase and merge it after this one is merged (I confirmed that it passes with this code).

bowlofeggs commented:
Let's see if Rawhide is working yet :)

bowlofeggs commented:
jenkies test

bowlofeggs commented:
well lookee there - rawhide works again. Thanks for the patch!
